nibabel/streamlines/array_sequence.py

```diff
@@ -0,0 +1,256 @@
+            lengths = list(map(len, elements))
+            self._lengths = np.r_[self._lengths, lengths]
+            self._offsets = np.r_[self._offsets,
+                                  np.cumsum([next_offset] + lengths)[:-1]]
+
+    def copy(self):
+        """ Creates a copy of this :class:`ArraySequence` object. """
+        # We do not simply deepcopy this object since we might have a chance
+        # to use less memory. For example, if the array sequence being copied
+        # is the result of a slicing operation on a array sequence.
+        clist = ArraySequence()
+        total_lengths = np.sum(self._lengths)
+        clist._data = np.empty((total_lengths,) + self._data.shape[1:],
+                               dtype=self._data.dtype)
+
+        next_offset = 0
+        offsets = []
+        for offset, length in zip(self._offsets, self._lengths):
+            offsets.append(next_offset)
+            chunk = self._data[offset:offset+length]
+            clist._data[next_offset:next_offset+length] = chunk
+            next_offset += length
+
+        clist._offsets = np.asarray(offsets)
+        clist._lengths = self._lengths.copy()
+
+        return clist
+
+    def __getitem__(self, idx):
+        """ Gets sequence(s) through advanced indexing.
+
+        Parameters
+        ----------
+        idx : int or slice or list or ndarray
+            If int, index of the element to retrieve.
+            If slice, use slicing to retrieve elements.
+            If list, indices of the elements to retrieve.
+            If ndarray with dtype int, indices of the elements to retrieve.
+            If ndarray with dtype bool, only retrieve selected elements.
+
+        Returns
+        -------
+        ndarray or :class:`ArraySequence`
+            If `idx` is an int, returns the selected sequence.
+            Otherwise, returns a :class:`ArraySequence` object which is view
+            of the selected sequences.
+
+        """
+        if isinstance(idx, int) or isinstance(idx, np.integer):
+            start = self._offsets[idx]
+            return self._data[start:start+self._lengths[idx]]
+
+        elif isinstance(idx, slice) or isinstance(idx, list):
+            clist = ArraySequence()
+            clist._data = self._data
+            clist._offsets = self._offsets[idx]
+            clist._lengths = self._lengths[idx]
+            return clist
+
+        elif isinstance(idx, np.ndarray) and np.issubdtype(idx.dtype,
+                                                           np.integer):
+            clist = ArraySequence()
+            clist._data = self._data
+            clist._offsets = self._offsets[idx]
+            clist._lengths = self._lengths[idx]
+            return clist
+
+        elif isinstance(idx, np.ndarray) and idx.dtype == np.bool:
+            clist = ArraySequence()
+            clist._data = self._data
+            clist._offsets = [self._offsets[i]
```

Comment:
Actually, I forgot to change those lines. `self._offsets` and `self._lengths` are now ndarrays, so we can do `self._offsets[idx]` and `self._lengths[idx]` which is the same as the `elif` just before. I'll merge the two cases.